[1.20.6] Add ComponentEnergyStorage for using IEnergyStorage with Data Components

Shadows-of-Fire opened the pull request:
This PR adds the new `ComponentEnergyStorage` class for ootb handling of `IEnergyStorage` with data components, since `EnergyStorage` is no longer suitable for this purpose.

Usage of the new class is fairly straightforward, create a `DataComponentType<Integer>` for your item, and then in your `ICapabilityProvider`, reference it to create an `ItemEnergyStorage` for that item.

The class will handle writing / reading to / from the itemstack as the data changes.

A future PR will also add a variant for `IItemHandler`.

Shadows-of-Fire committed:
Add ItemEnergyStorage

Shadows-of-Fire committed:
Add simple usage docs

Technici4n commented on src/main/java/net/neoforged/neoforge/energy/ItemEnergyStorage.java:
```diff
@@ -0,0 +1,113 @@
+            return 0;
+
+        int energy = this.getEnergyStored();
+        int energyReceived = Math.min(capacity - energy, Math.min(this.maxReceive, maxReceive));
```
If capacity < energy this will return a negative value (bad). It should return 0 if that happens, as it is possible when configs change etc...

Technici4n commented on src/main/java/net/neoforged/neoforge/energy/ItemEnergyStorage.java:
```diff
@@ -0,0 +1,113 @@
+
+    @Override
+    public int receiveEnergy(int maxReceive, boolean simulate) {
+        if (!canReceive())
```
Do we want to return early for maxReceive <= 0?

Shadows-of-Fire replied:
Well, yes, that's what this condition does

Technici4n replied:
The parameter I mean.

Shadows-of-Fire replied:
Yeah, updated (both `ItemEnergyStorage` and `EnergyStorage`) to have the short-circuiting behavior, as well as renaming the param to `toReceive` / `toExtract` to reduce confusion with the field.

Technici4n commented on src/main/java/net/neoforged/neoforge/energy/ItemEnergyStorage.java:
```diff
@@ -0,0 +1,113 @@
+    @Override
+    public int getEnergyStored() {
+        int rawEnergy = this.parent.getOrDefault(this.powerComponent, 0);
+        return Math.max(0, Math.min(capacity, rawEnergy));
```
Does Mth have a clamp method for this?

Technici4n commented on src/main/java/net/neoforged/neoforge/energy/ItemEnergyStorage.java:
```diff
@@ -0,0 +1,113 @@
+        return this.maxReceive > 0;
+    }
+
+    protected void setEnergy(int energy) {
```
Do we want to delete the component when it goes down to 0? Maybe not.

Shadows-of-Fire replied:
Probably not, no.  I think that would be confusing against typical user expectations (and is probably in contrast of how components are normally used)

Shadows-of-Fire committed:
Use Mth.clamp, rename params, and short-circuit on negatives

Renames maxReceive and maxExtract params to toReceive and toExtract respectively, reducing confusion with the fields of the same name.

Shadows-of-Fire committed:
Rename to `ComponentEnergyStorage`

Shadows-of-Fire commented:
Updated to rename `ItemEnergyStorage` to `ComponentEnergyStorage` for naming consistency with the planned `ComponentItemHandler` class.

pupnewfster commented on src/main/java/net/neoforged/neoforge/energy/ComponentEnergyStorage.java:
```diff
@@ -0,0 +1,123 @@
+        int energy = this.getEnergyStored();
+        int energyReceived = Mth.clamp(this.capacity - energy, 0, Math.min(this.maxReceive, toReceive));
+        if (!simulate) {
+            this.setEnergy(energy + energyReceived);
```
If energyReceived is zero, because it is already full we want to skip bothering with the component mutation via setEnergy. And likewise for extractEnergy if it is already empty

Shadows-of-Fire committed:
Avoid component writes if nothing changed

Technici4n commented on src/main/java/net/neoforged/neoforge/energy/ComponentEnergyStorage.java:
```diff
@@ -0,0 +1,123 @@
+ */
+public class ComponentEnergyStorage implements IEnergyStorage {
+    protected final MutableDataComponentHolder parent;
+    protected final DataComponentType<Integer> energyComponent;
```
Do we want to change this to a `Supplier<DataComponentType<Integer>>`?

Shadows-of-Fire replied:
I don't think there's any benefit in doing that, it just moves the .get() call from the consumer (at the ctor) to every field access

Technici4n replied:
Just asking for consistency with `FluidHandlerItemStack`.

Shadows-of-Fire replied:
Electing to proceed with the non-supplier options per https://discord.com/channels/313125603924639766/1154167065519861831/1244080298795335721

It wouldn't make sense to use the supplier when we need to immediately resolve it for sanity checking.
